Source/Scene/ModelExperimental/MaterialPipelineStage.js

```diff
@@ -0,0 +1,449 @@
+    ShaderDestination.FRAGMENT
+  );
+  uniformMap[transformUniformName] = function () {
+    return textureReader.transform;
+  };
+}
+
+/**
+ * Process a single texture and add it to the shader and uniform map.
+ *
+ * @param {ShaderBuilder} shaderBuilder The shader builder to modify
+ * @param {Object.<String, Function>} uniformMap The uniform map to modify.
+ * @param {ModelComponents.TextureReader} textureReader The texture to add to the shader
+ * @param {String} uniformName The name of the sampler uniform such as <code>u_baseColorTexture</code>
+ * @param {String} defineName The name of the texture for use in the defines, minus any prefix or suffix. For example, "BASE_COLOR" or "EMISSIVE"
+ *
+ * @private
+ */
+function processTexture(
+  shaderBuilder,
+  uniformMap,
+  textureReader,
+  uniformName,
+  defineName,
+  defaultTexture
+) {
+  // Add a uniform for the texture itself
+  shaderBuilder.addUniform(
+    "sampler2D",
+    uniformName,
+    ShaderDestination.FRAGMENT
+  );
+  uniformMap[uniformName] = function () {
+    return defaultValue(textureReader.texture, defaultTexture);
+  };
+
+  // Add a #define directive to enable using the texture in the shader
+  var textureDefine = "HAS_" + defineName + "_TEXTURE";
+  shaderBuilder.addDefine(textureDefine, undefined, ShaderDestination.FRAGMENT);
+
+  // Add a #define to tell the shader which texture coordinates varying to use.
+  var texCoordIndex = textureReader.texCoord;
+  var texCoordVarying = "v_texCoord_" + texCoordIndex;
+  var texCoordDefine = "TEXCOORD_" + defineName;
+  shaderBuilder.addDefine(
+    texCoordDefine,
+    texCoordVarying,
+    ShaderDestination.FRAGMENT
+  );
+
+  // Some textures have matrix transforms (e.g. for texture atlases). Add those
+  // to the shader if present.
+  if (defined(textureReader.transform)) {
+    processTextureTransform(
+      shaderBuilder,
+      uniformMap,
+      textureReader,
+      uniformName,
+      defineName
+    );
+  }
+}
+
+function processMaterialUniforms(
+  material,
+  uniformMap,
+  shaderBuilder,
+  defaultTexture
+) {
+  var emissiveTexture = material.emissiveTexture;
+  if (defined(emissiveTexture)) {
+    processTexture(
+      shaderBuilder,
+      uniformMap,
+      emissiveTexture,
+      "u_emissiveTexture",
+      "EMISSIVE",
+      defaultTexture
+    );
+  }
+
+  var emissiveFactor = material.emissiveFactor;
+  if (defined(emissiveFactor)) {
+    shaderBuilder.addUniform(
+      "vec3",
+      "u_emissiveFactor",
+      ShaderDestination.FRAGMENT
+    );
+    uniformMap.u_emissiveFactor = function () {
+      return material.emissiveFactor;
+    };
+    shaderBuilder.addDefine(
+      "HAS_EMISSIVE_FACTOR",
+      undefined,
+      ShaderDestination.FRAGMENT
+    );
+  }
+
+  var normalTexture = material.normalTexture;
+  if (defined(normalTexture)) {
+    processTexture(
+      shaderBuilder,
+      uniformMap,
+      normalTexture,
+      "u_normalTexture",
+      "NORMAL",
+      defaultTexture
```

Comment:
I forget where we landed for the default texture for normal maps. Is the white color modified to be (0.5, 0.5, 1.0) in the shader?